File: modules/_foundation/src/main/java/com/projecthub/base/workflow/domain/service/WorkflowEngine.java
package com.projecthub.base.workflow.domain.service;

import com.projecthub.base.workflow.domain.model.*;
import com.projecthub.base.workflow.domain.event.WorkflowTransitionEvent;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class WorkflowEngine {
    private final ApplicationEventPublisher eventPublisher;

    /**
     * Attempts to transition an entity to a new state using the provided workflow definition
     * @throws IllegalStateException if the transition is not allowed or validation fails
     */
    @Transactional
    public void transition(WorkflowDefinition workflow, WorkflowContext context) {
        WorkflowState currentState = context.getCurrentState();
        WorkflowState targetState = context.getTargetState();

        // Verify the transition is allowed by the workflow definition
        if (!workflow.isTransitionAllowed(currentState, targetState)) {
            throw new IllegalStateException(
                String.format("Transition from %s to %s is not allowed in workflow %s",
                    currentState.getName(), targetState.getName(), workflow.getName())
            );
        }

        // Find the transition
        WorkflowTransition transition = workflow.getAvailableTransitions(currentState).stream()
            .filter(t -> t.getToState().equals(targetState))
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("Transition not found"));

        // Validate the transition
        if (!transition.isValid(context)) {
            throw new IllegalStateException("Transition validation failed");
        }

        // Execute transition actions
        transition.executeActions(context);

        // Publish transition event
        publishTransitionEvent(workflow, context, transition);
    }

    private void publishTransitionEvent(WorkflowDefinition workflow, WorkflowContext context, WorkflowTransition transition) {
        WorkflowTransitionEvent event = WorkflowTransitionEvent.builder()
            .workflowId(workflow.getId())
            .entityId(context.getEntityId())
            .entityType(context.getEntityType())
            .fromState(transition.getFromState())
            .toState(transition.getToState())
            .transitionName(transition.getName())
            .build();

        eventPublisher.publishEvent(event);
    }
}

File: src/main/kotlin/com/projecthub/application/workflow/service/WorkflowEngine.kt
package com.projecthub.core.application.service

import com.projecthub.core.domain.model.workflow.*
import com.projecthub.core.domain.event.WorkflowTransitionEvent
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class WorkflowEngine(private val eventPublisher: ApplicationEventPublisher) {

    @Transactional
    fun transition(workflow: WorkflowDefinition, context: WorkflowContext) {
        val currentState = context.getCurrentState()
        val targetState = context.getTargetState()

        if (!workflow.isTransitionAllowed(currentState, targetState)) {
            throw IllegalStateException(
                "Transition from ${currentState.getName()} to ${targetState.getName()} is not allowed in workflow ${workflow.getName()}"
            )
        }

        val transition = workflow.getAvailableTransitions(currentState).stream()
            .filter { t -> t.getToState() == targetState }
            .findFirst()
            .orElseThrow { IllegalStateException("Transition not found") }

        if (!transition.isValid(context)) {
            throw IllegalStateException("Transition validation failed")
        }

        transition.executeActions(context)
        publishTransitionEvent(workflow, context, transition)
    }

    private fun publishTransitionEvent(
        workflow: WorkflowDefinition,
        context: WorkflowContext,
        transition: WorkflowTransition
    ) {
        val event = WorkflowTransitionEvent.builder()
            .workflowId(workflow.getId())
            .entityId(context.getEntityId())
            .entityType(context.getEntityType())
            .fromState(transition.getFromState())
            .toState(transition.getToState())
            .transitionName(transition.getName())
            .build()

        eventPublisher.publishEvent(event)
    }
}
